Ignore single files with `@noformat` / `@noprettier`

wnayes opened the pull request:
## Description

This is an implementation of `@noformat` / `@noprettier` support as discussed in #3634.

Languages that had existing support for `@format` parsing were updated to support this new feature:
- JavaScript (TypeScript)
- CSS
- HTML
- JSON
- Markdown (MDX)
- YAML
- GraphQL

Examples:
#### JavaScript
```jsx
/**
 * @noformat
 */

export default matrix(
  1, 0, 0,
  0, 1, 0,
  0, 0, 1
);
```

#### CSS
```css
/** @noprettier -- automatically generated file */
.class { display: none; }
```

-----

### Rationale

This creates a way to opt out individual files from formatting in a decentralized manner.

I am considering approaches to enforce Prettier on a large repository with many developers. The current `requirePragma`/`insertPragma` approach could work, but it seems like an uphill battle to enforce that everyone add `@format` or similar all new files going forward. Having something like `@noformat` would allow marking an initial set of opted out files, and these pragmas could be removed over time as Prettier is adopted.

`.prettierignore` can also be used today to ignore a set of files, but I don't expect it to scale as well as this solution when the number of opted out files is initially very large.

### Implementation

The core feature here is to let plugins implement a new `hasIgnorePragma` function, similar to the existing `hasPragma`. In `hasIgnorePragma`, they check for `@noformat` or `@noprettier` instead. The core calls this around the same time as the requirePragma check.

Currently, I gated this behavior behind a new option, `checkIgnorePragma`. My thinking here was mainly around performance; it seems not great to subject everyone to an extra docblock check during formatting if they don't intend to use the feature. The existence of this new option, as well as its default (false), could certainly be reconsidered.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

fisker commented on docs/options.md:
```diff
@@ -393,6 +393,14 @@ Note that “in tandem” doesn’t mean “at the same time”. When the two op
 
+## Check Ignore Pragma
+
+Whether or not to detect and respect `@noprettier` or `@noformat` markers at the top of files. These will prevent formatting when this option is enabled.
```
We should mention the reason why this is disabled by default.

> My thinking here was mainly around performance

wnayes replied:
Rewrote this options.md section to include the rationale.

fisker commented:
I like this.

wnayes commented:
Pushed revisions and additional commits to build out the feature more fully. Added support + tests to the languages that had existing pragma parsing. Added a draft changelog entry.

fisker commented on src/main/core.js:
```diff
@@ -311,6 +311,14 @@ async function hasPragma(text, options) {
 
+async function hasIgnorePragma(text, options) {
+  const selectedParser = await resolveParser(options);
+  return (
+    Boolean(selectedParser.hasIgnorePragma) &&
+    selectedParser.hasIgnorePragma(text)
+  );
```
```suggestion
  return selectedParser.hasIgnorePragma?.(text);
```

wnayes replied:
Made this change.

fisker commented on src/language-markdown/pragma.js:
```diff
@@ -1,11 +1,11 @@
 import parseFrontMatter from "../utils/front-matter/parse.js";
 
-const pragmas = ["format", "prettier"];
+const FormatPragmas = ["format", "prettier"];
+const NoFormatPragmas = ["noformat", "noprettier"];
```
```suggestion
const FORMAT_PRAGMAS = ["format", "prettier"];
const NO_FORMAT_PRAGMAS = FORMAT_PRAGMAS.map(pragma => `no-${pragma}`);
```

Maybe we should put this constants somewhere so all plugins can use it.

fisker replied:
Done in #17346, please rebase.

wnayes replied:
Thanks, I rebased and tried to follow this pattern. I may not have understood the intent with the FIXME comments that were added; since I just added additional ignore code, I removed those FIXMEs.

fisker replied:
> the FIXME comments that were added

Sorry, mistake. Thanks for fixing them.

fisker commented on src/main/core.js:
```diff
@@ -319,7 +327,8 @@ async function formatWithCursor(originalText, originalOptions) {
     (options.rangeStart >= options.rangeEnd && text !== "") ||
-    (options.requirePragma && !(await hasPragma(text, options)))
+    (options.requirePragma && !(await hasPragma(text, options))) ||
+    (options.checkIgnorePragma && (await hasIgnorePragma(text, options)))
```
1. Should we enable this feature if `--require-pragma` used?
2. What will happen if two flags are used together?

wnayes replied:
Currently, the two options act independently and look at different pragmas.
- If you have `requirePragma` enabled, you need `@prettier` in order to format.
- If you have `checkIgnorePragma enabled, you need to not have `@noprettier` to format.

If both options are enabled, you only format files that have `@prettier` and don't have `@noprettier`. From the perspective of both features limiting which files get formatted, neither really has power over the other; both can exclude files for different reasons.

That's my take on it anyway, I don't feel too strongly. That's the way it landed in implementing it, and other behaviors would likely require additional code.

I could add tests if we want to formalize the behavior.

fisker commented:
What will happen if `@prettier` and `@noprettier` both present?

fisker reviewed (approved):
LGTM, thanks!

fisker commented on changelog_unreleased/api/17344.md:
```diff
@@ -0,0 +1,20 @@
+#### Ignore files with `--check-ignore-pragma` (#17344 by @wnayes)
```
The PR tile

> Ignore single files with `@noformat` / `@noprettier`

is fine as a commit title.

But it feels the feature is supported out-of-box (requires an extra flag to work, actually), so I updated to this version.

fisker commented:
I'll look into the test failue.

wnayes commented:
Thanks for looking further into this. I pushed a couple commits with small doc wording changes, but otherwise I have no further changes.

fisker commented:
> I pushed a couple commits with small doc wording changes,

I changed them to align with words in `requirePragma`, please make sure they are aligned.

sosukesuzuki commented on changelog_unreleased/api/17344.md:
```diff
@@ -0,0 +1,20 @@
+#### Ignore files with `--check-ignore-pragma` (#17344 by @wnayes)
+
+Individual files can now opt out of formatting via `@noformat` or `@noprettier` "pragma" comments at the top of the file.
```
Why we should have both `@noformat` and `@noprettier` ? I feel one would be enough.
(I personally prefer `@noprettier`)

wnayes replied:
One way I've been thinking about it is
- `@noprettier` means Prettier specifically should not be used.
- `@noformat` means no kind of formatter should be used (of which Prettier is one).

They each can express different sort of intents. Other formatters would have to align on the `@noformat` concept for that to be effective, though.

The other practical factor here is simply that `requirePragma` had supported `@prettier` and `@format`, so this reflects that.

sosukesuzuki replied:
> The other practical factor here is simply that requirePragma had supported @prettier and @format, so this reflects that.

Ah it makes sense to align it with `requirePragma`.

sosukesuzuki reviewed (commented):
Overall LGTM!! I like this feature, thank you

sosukesuzuki reviewed (approved):
LGTM

sosukesuzuki commented:
@fisker I've approved this. Please make minor tweaks if necessary and please merge this!

romgrk commented:
Why is the `/**` format required though? In files without imports, it attaches a jsdoc comment on an unrelated statement.

![image](https://github.com/user-attachments/assets/2dac527e-9bce-4d61-9657-397a0331c6aa)
